test/jdk/sun/security/provider/hss/TestHSSLMS.java

```diff
@@ -0,0 +1,2597 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug JDK-8298127
+ * @library /test/lib
+ * @summary tests for HSS/LMS provider
+ * @modules java.base/sun.security.util
+ * @run main TestHSSLMS
+ */
+
+import java.io.*;
+import java.security.KeyFactory;
+import java.security.PublicKey;
+import java.security.Security;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.HexFormat;
+
+import sun.security.util.*;
+
+import jdk.test.lib.util.SerializationUtils;
+
+public class TestHSSLMS {
+    static final String ALG = "HSS/LMS";
+    static final String OID = "1.2.840.113549.1.9.16.3.17";
+
+    public static void main(String[] args) throws Exception {
+
+        // Tests 6-13 were generated with Bouncy Castle using parameter sets
+        // mentioned in RFC 8554 section 6.4: two with W=8 and six with W=4.
+
+        int i = 1;
+        for (TestCase t : TestCases) {
+            if (!kat(t)) {
+                throw new RuntimeException("test case #" + i + " failed");
+            }
+            i++;
+        }
+
+        if (!serializeTest()) {
+            throw new RuntimeException("serializeTest failed");
+        }
+
+        System.out.println("All tests passed");
+    }
+
+    static boolean kat(TestCase t) throws Exception {
+        if (!t.exception) {
+            if (verify(t.pk, t.sig, t.msg)) {
+                return t.expected;
+            } else {
+                return !t.expected;
+            }
+        } else {
+            // exception is expected and is a pass (true)
+            try {
+                verify(t.pk, t.sig, t.msg);
+                return false;
+            } catch (InvalidKeySpecException | SignatureException ex) {
```

Comment:
It would be nice to test for the expected exception, either `InvalidKeySpecException` or `SignatureException`. You could add an additional field to the `TestCase` record which checks for the expected exception type.

Reply:
Fixed.